src/apply.rs

```diff
@@ -49,35 +58,114 @@ impl ApplyStats {
             lines_deleted: 0,
             lines_context: 0,
             hunks_applied: 0,
-            has_changes: false,
         }
     }
 
     /// Add statistics from a hunk
-    fn add_hunk(&mut self, added: usize, deleted: usize, context: usize) {
-        self.lines_added += added;
-        self.lines_deleted += deleted;
-        self.lines_context += context;
+    fn add_hunk(&mut self, hunk_stats: HunkStats) {
+        self.lines_added += hunk_stats.added;
+        self.lines_deleted += hunk_stats.deleted;
+        self.lines_context += hunk_stats.context;
         self.hunks_applied += 1;
-        if added > 0 || deleted > 0 {
-            self.has_changes = true;
-        }
+    }
+
+    /// Returns whether any changes were made (false if patch was already applied)
+    pub fn has_changes(&self) -> bool {
+        self.lines_added > 0 || self.lines_deleted > 0
     }
 }
 
 /// Result of applying a patch with statistics
+///
+/// Similar to `Result<T, E>` but includes statistics about the patch application
+/// even on success.
 #[derive(Clone, Debug)]
-pub struct ApplyResult<T> {
-    /// The patched content
-    pub content: T,
-    /// Statistics about the changes made
-    pub stats: ApplyStats,
+pub enum ApplyResult<T, E> {
```

Comment:
I think it would make more sense to do type alias like this https://docs.rs/nom/latest/nom/type.IResult.html instead of reimplementing Result type from scratch.